### PR TITLE
Allow custom index.html template registration

### PR DESCRIPTION
**Pull request by ge3224**
This PR addresses [Issue #5](https://github.com/olivere/vite/issues/5), allowing for custom `index.html` template registration without causing a panic. The key changes include:

1. **Custom index.html Registration**: Modified the behavior so that the default index.html template is now a fallback (renamed as fallback.html). This allows users to register their custom index.html templates without encountering a panic.
3. **Improved Template Lookup**: Enhanced the template lookup process to attempt finding alternative template names if the exact match isn't found, reducing the risk of errors due to naming mismatches.
4. **Example Directory**: Added a new example in the `examples/template-registration` directory demonstrating how to register and use custom templates.

Feel free to review the changes and provide feedback. Let me know if there are any additional adjustments needed

### CODE REVIEW

File: handler.go [hunk 1]
```diff
@@ -244,11 +252,42 @@ func (h *Handler) renderPage(w http.ResponseWriter, r *http.Request, path string
...
+		tmplName = "index.html"
+	} else {
+		tmplName = path
+	}
+
 	// Find the template to use.
-	tmpl, ok := h.templates[path]
+	tmpl, ok := h.templates[tmplName]
+	if !ok {
+		// Try alternative names
```
Comment (olivere) [inline handler.go:265]: Maybe we can add a comment about what this actually does? I mean, yes, anyone can scan it. But it takes a few seconds to read, and a comment just makes it obvious with a glimpse. We could even extract this to an individual method (but this seems optional to me). What do you think?

File: handler.go [hunk 2]
```diff
@@ -244,11 +252,42 @@ func (h *Handler) renderPage(w http.ResponseWriter, r *http.Request, path string
...
 	if !ok {
-		// Use index.html by default
-		tmpl = h.templates["index.html"]
+		// check if custom templates were registered, perhaps a mistake was made in naming
+		if len(h.templates) > 1 {
+			keys := make([]string, 0, len(h.templates))
+			for k := range h.templates {
+				keys = append(keys, k)
+			}
+			log.Printf("Warning: template %q not found. Available templates: %s", tmplName, strings.Join(keys, ", "))
```
Comment (olivere) [inline handler.go:288]: Hmm... How do libraries log warnings nowadays? Do we use `log` or `slog.Warn(...)`, or invent a logger interface (which I don't like)?

File: handler.go [hunk 3]
```diff
@@ -269,7 +310,11 @@ var (
 	{{- if .IsDev }}
 		{{ .PluginReactPreamble }}
 		<script type="module" src="{{ .ViteURL }}/@vite/client"></script>
-		<script type="module" src="{{ .ViteURL }}/src/main.tsx"></script>
+		{{ if ne .ViteEntry "" }}
```
Comment (olivere) [inline handler.go:313]: Maybe replace this with ...

```suggestion
		{{- if ne .ViteEntry "" }}
```

?

File: handler.go [hunk 4]
```diff
@@ -269,7 +310,11 @@ var (
...
+			<script type="module" src="{{ .ViteURL }}/{{ .ViteEntry }}"></script>
+		{{ else }}
```
Comment (olivere) [inline handler.go:315]: And here ...

```suggestion
		{{- else }}
```

File: handler.go [hunk 5]
```diff
@@ -269,7 +310,11 @@ var (
...
+			<script type="module" src="{{ .ViteURL }}/src/main.tsx"></script>
+		{{ end }}
```
Comment (olivere) [inline handler.go:317]: ... and finally:

```suggestion
		{{- end }}
```

... as that would make the output a bit cleaner?

**Review by olivere — CHANGES_REQUESTED** [review]
LGTM overall. Can you have a look at my comments, if you have the time?

File: handler.go (hunk 1 above)
Comment (ge3224) [reply handler.go:265]: That's good. What do you think of something like this?

```go
	// Find the template by name.
	tmpl, ok := h.templates[tmplName]

	// Catch common variations. If a template isn't found by the exact name,
	// check for variations like: "page", "page.html", or "/page.html", to match
	// how users might have registered the template.
	if !ok {
		variations := []string{
			strings.TrimPrefix(tmplName, "/"),
			strings.TrimPrefix(tmplName, "/") + ".html",
			strings.TrimSuffix(strings.TrimPrefix(tmplName, "/"), ".html"),
			tmplName + ".html",
		}
		for _, variant := range variations {
			if t, found := h.templates[variant]; found {
				tmpl = t
				ok = true
				break
			}
		}
	}
```

File: handler.go (hunk 2 above)
Comment (ge3224) [reply handler.go:288]: Thanks. I've been out of Go for a while and wasn't aware of the structured logging package. What do you think of something like this?

```go
	// Handle case when requested template is not found:
	// 1. If multiple templates exist, log a warning with the requested and available templates.
	// 2. Fall back to a default template.
	if !ok {
		if len(h.templates) > 1 {
			keys := make([]string, 0, len(h.templates))
			for k := range h.templates {
				keys = append(keys, k)
			}
			slog.Warn(
				"Template not found",
				"requestedTemplate", tmplName,
				"availableTemplates", strings.Join(keys, ", "),
			)
		}
		tmpl = h.templates[fallbackTemplateName]
	}
```

File: handler.go (hunk 3 above)
Comment (ge3224) [reply handler.go:313]: Yes. Will do.

File: handler.go (hunk 4 above)
Comment (ge3224) [reply handler.go:315]: Done.

File: handler.go (hunk 5 above)
Comment (ge3224) [reply handler.go:317]: Will do.

**Comment by ge3224** [comment]
In case you haven't seen it yet, I've pushed the requested changes to this PR. Feel free to take another look when convenient. Happy to make any further adjustments if needed.

**Review by olivere — APPROVED** [review]
LGTM. Will do a final review and merge then. Thanks a lot.

**Pull request merged**